Jax fixes

zimbatm opened the pull request:
## Description of changes

Doing a pass to get jax are related libraries to a good place.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

zimbatm committed:
python3Packages.jaxlib: switch version based on support

So that packages depending on jaxlib can also build on macOS.

zimbatm committed:
python3Packages.jax: disable segfaulting test

mweinelt commented:
Deferring to the CUDA team, they are much closer to Jax.

SomeoneSerge commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6114,13 +6114,13 @@ self: super: with self; {
   };
 
-  jaxlib = self.jaxlib-build;
+  jaxlib = if self.jaxlib-build.meta.broken then self.jaxlib-bin else self.jaxlib-build;
```
Ohhh I'm not sure how good of an idea this is: I can see myself tearing my hair out a few steps down the line (steps involving changing meta.unfree) trying to figure out how I ended up with an unfree pypi wheel in the closure instead of a cpu-only source-build

GaetanLepage replied:
I personally don't mind, but I get your point. Feel free to decide.

samuela replied:
I agree that this feels like a footgun. 

Another option: we alias `jaxlib = jaxlib-bin` and leave a comment linking to this PR and the hash mismatch issue.

zimbatm replied:
Ok, I like samuela's simple approach. Fixed.

SomeoneSerge replied:
I'd suggest a more radical approach, which would probably cause frustration to other maintainers: we add a `lib.warn` to the `if broken then jaxlib-bin` branch saying that we're using a prebuilt and patchelfed jaxlib from pypi, potentially including unfree components

SomeoneSerge replied:
It'd be annoying because it'll show warnings in the default nixpkgs instantiation and in ofborg

SomeoneSerge replied:
And to be clear, the only admissible reason I see to add the alias is that this will get hydra and ofborg build reverse dependencies. Otherwise people could do this in a local overlay and we probably should do a better job advertising how to do overlays

SomeoneSerge replied:
> The "if broken then jaxlib-bin else jaxlib-build" logic existed in the jax package.

Ohhh, I had completely missed that

SomeoneSerge replied:
Ok, there was already an alias in place, and in this PR it place two roles:

- Supporting darwin
- Working around the hash issue

For the former, I'd replace the logic with `if jaxlib.meta.unsupported then jaxlib-bin else jaxlib` to look at `platforms` instead of `broken`. For the latter, I hope we can work around the instability soon.

> And if the build version is broken, the packages should keep working.

Yes but whether shipping a substitute "product" counts as "it's still working" is a matter of setting expectations.

Anyway I see now that my comment was out of scope

samuela replied:
> The "if broken then jaxlib-bin else jaxlib-build" logic existed in the jax package.

Yes, I wrote this in order to fix the build on darwin.

> Ok, there was already an alias in place

Actually, there is/was currently no such alias in place. The "if broken then jaxlib-bin else jaxlib-build" was only logic local to the jax package so that it would use jaxlib in `checkPhase`, but its impact was limited to that. See https://github.com/NixOS/nixpkgs/blob/6a8b9742401bba8771e5f8afe7b240b4410869cd/pkgs/development/python-modules/jax/default.nix#L29

SuperSandro2000 replied:
I think just writting ``if stdenv.isDarwin then A else B`` is way easier to understand and we should probably just go with that.

zimbatm commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -77,7 +77,9 @@ let
     # NOTE: We always build with NCCL; if it is unsupported, then our build is broken.
-    broken = effectiveStdenv.isDarwin || nccl.meta.unsupported;
+
+    # FIXME: Bazel fetch is not reproducible
```
fixed

zimbatm commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -130,10 +127,14 @@ buildPythonPackage rec {
 
-  disabledTestPaths = lib.optionals (stdenv.isDarwin && stdenv.isAarch64) [
+  disabledTestPaths = [
+    # Segmentation fault
```
Not yet. I don't know if this is an upstream or a packaging issue.

zimbatm replied:
```
python3.11-jax> sssssssssssssssssssssssssssss.Fatal Python error: Segmentation fault
python3.11-jax> Thread 0x00007ffff6bff6c0 (most recent call first):
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 534 in read
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 567 in from_io
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 1160 in _thread_receiver
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 341 in run
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 411 in _perform_spawn
python3.11-jax> Thread 0x00007ffff7eae740 (most recent call first):
python3.11-jax>   File "/build/source/jax/_src/interpreters/pxla.py", line 1185 in __call__
python3.11-jax>   File "/build/source/jax/_src/profiler.py", line 335 in wrapper
python3.11-jax>   File "/build/source/jax/_src/pjit.py", line 1462 in _pjit_call_impl_python
python3.11-jax>   File "/build/source/jax/_src/pjit.py", line 1508 in call_impl_cache_miss
python3.11-jax>   File "/build/source/jax/_src/pjit.py", line 1525 in _pjit_call_impl
python3.11-jax>   File "/build/source/jax/_src/core.py", line 879 in process_primitive
python3.11-jax>   File "/build/source/jax/_src/core.py", line 391 in bind_with_trace
python3.11-jax>   File "/build/source/jax/_src/core.py", line 2789 in bind
python3.11-jax>   File "/build/source/jax/_src/pjit.py", line 1745 in _pjit_batcher
python3.11-jax>   File "/build/source/jax/_src/interpreters/batching.py", line 433 in process_primitive
python3.11-jax>   File "/build/source/jax/_src/core.py", line 391 in bind_with_trace
python3.11-jax>   File "/build/source/jax/_src/core.py", line 2789 in bind
python3.11-jax>   File "/build/source/jax/_src/pjit.py", line 181 in _python_pjit_helper
python3.11-jax>   File "/build/source/jax/_src/pjit.py", line 304 in cache_miss
python3.11-jax>   File "/build/source/jax/_src/traceback_util.py", line 179 in reraise_with_filtered_traceback
python3.11-jax>   File "/build/source/jax/_src/scipy/linalg.py", line 477 in eigh
python3.11-jax>   File "/build/source/jax/_src/linear_util.py", line 192 in call_wrapped
python3.11-jax>   File "/build/source/jax/_src/api.py", line 1214 in vmap_f
python3.11-jax>   File "/build/source/jax/_src/traceback_util.py", line 179 in reraise_with_filtered_traceback
python3.11-jax>   File "/build/source/tests/linalg_test.py", line 568 in testEighBatching
python3.11-jax>   File "/nix/store/bz8980qcci1rzicwg8ll0i8k0zqk5y1b-python3.11-absl-py-2.1.0/lib/python3.11/site-packages/absl/testing/parameterized.py", line 319 in bound_param_test
python3.11-jax>   File "/nix/store/6b1fqdwb3g56j5pazv8zkx9qd0mv3wiz-python3-3.11.9/lib/python3.11/unittest/case.py", line 579 in _callTestMethod
python3.11-jax>   File "/nix/store/6b1fqdwb3g56j5pazv8zkx9qd0mv3wiz-python3-3.11.9/lib/python3.11/unittest/case.py", line 623 in run
python3.11-jax>   File "/nix/store/6b1fqdwb3g56j5pazv8zkx9qd0mv3wiz-python3-3.11.9/lib/python3.11/unittest/case.py", line 678 in __call__
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/unittest.py", line 321 in runtest
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/runner.py", line 172 in pytest_runtest_call
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_callers.py", line 102 in _multicall
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_manager.py", line 119 in _hookexec
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_hooks.py", line 501 in __call__
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/runner.py", line 240 in <lambda>
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/runner.py", line 340 in from_call
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/runner.py", line 239 in call_and_report
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/runner.py", line 134 in runtestprotocol
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/runner.py", line 115 in pytest_runtest_protocol
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_callers.py", line 102 in _multicall
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_manager.py", line 119 in _hookexec
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_hooks.py", line 501 in __call__
python3.11-jax>   File "/nix/store/hcb91kq4vchiax4ljvj4nj0azs5hx6d9-python3.11-pytest-xdist-3.5.0/lib/python3.11/site-packages/xdist/remote.py", line 174 in run_one_test
python3.11-jax>   File "/nix/store/hcb91kq4vchiax4ljvj4nj0azs5hx6d9-python3.11-pytest-xdist-3.5.0/lib/python3.11/site-packages/xdist/remote.py", line 157 in pytest_runtestloop
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_callers.py", line 102 in _multicall
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_manager.py", line 119 in _hookexec
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_hooks.py", line 501 in __call__
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/main.py", line 339 in _main
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/main.py", line 285 in wrap_session
python3.11-jax>   File "/nix/store/8qyy6cj0w77g2n6rjp674fv7sjxm4xw1-python3.11-pytest-8.1.1/lib/python3.11/site-packages/_pytest/main.py", line 332 in pytest_cmdline_main
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_callers.py", line 102 in _multicall
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_manager.py", line 119 in _hookexec
python3.11-jax>   File "/nix/store/4rbwjxqk4vh2k2rr7cmdv09rfka5bq0p-python3.11-pluggy-1.4.0/lib/python3.11/site-packages/pluggy/_hooks.py", line 501 in __call__
python3.11-jax>   File "/nix/store/hcb91kq4vchiax4ljvj4nj0azs5hx6d9-python3.11-pytest-xdist-3.5.0/lib/python3.11/site-packages/xdist/remote.py", line 355 in <module>
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 1291 in executetask
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 341 in run
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 411 in _perform_spawn
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 389 in integrate_as_primary_thread
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 1273 in serve
python3.11-jax>   File "/nix/store/4s1y6idpq695dyja9br66yhcfnpp75zw-python3.11-execnet-2.1.1/lib/python3.11/site-packages/execnet/gateway_base.py", line 1806 in serve
python3.11-jax>   File "<string>", line 8 in <module>
python3.11-jax>   File "<string>", line 1 in <module>
python3.11-jax> Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, jaxlib.cpu_feature_guard, PIL._imaging, scipy._lib._ccallback_c, scipy.sparse._sparsetools, _csparsetools, scipy.sparse._csparsetools, scipy.linalg._fblas, scipy.linalg._flapack, scipy.linalg.cython_lapack, scipy.linalg._cythonized_array_utils, scipy.linalg._solve_toeplitz, scipy.linalg._decomp_lu_cython, scipy.linalg._matfuncs_sqrtm_triu, scipy.linalg.cython_blas, scipy.linalg._matfuncs_expm, scipy.linalg._decomp_update, scipy.sparse.linalg._dsolve._superlu, scipy.sparse.linalg._eigen.arpack._arpack, scipy.sparse.linalg._propack._spropack, scipy.sparse.linalg._propack._dpropack, scipy.sparse.linalg._propack._cpropack, scipy.sparse.linalg._propack._zpropack, scipy.sparse.csgraph._tools, scipy.sparse.csgraph._shortest_path, scipy.sparse.csgraph._traversal, scipy.sparse.csgraph._min_spanning_tree, scipy.sparse.csgraph._flow, scipy.sparse.csgraph._matching, scipy.sparse.csgraph._reordering, scipy.special._ufuncs_cxx, scipy.special._cdflib, scipy.special._ufuncs, scipy.special._specfun, scipy.special._comb, scipy.special._ellip_harm_2, scipy.integrate._odepack, scipy.integrate._quadpack, scipy.integrate._vode, scipy.integrate._dop, scipy.integrate._lsoda, scipy.optimize._minpack2, scipy.optimize._group_columns, scipy._lib.messagestream, scipy.optimize._trlib._trlib, scipy.optimize._lbfgsb, _moduleTNC, scipy.optimize._moduleTNC, scipy.optimize._cobyla, scipy.optimize._slsqp, scipy.optimize._minpack, scipy.optimize._lsq.givens_elimination, scipy.optimize._zeros, scipy.optimize._highs.cython.src._highs_wrapper, scipy.optimize._highs._highs_wrapper, scipy.optimize._highs.cython.src._highs_constants, scipy.optimize._highs._highs_constants, scipy.linalg._interpolative, scipy.optimize._bglu_dense, scipy.optimize._lsap, scipy.spatial._ckdtree, scipy.spatial._qhull, scipy.spatial._voronoi, scipy.spatial._distance_wrap, scipy.spatial._hausdorff, scipy.spatial.transform._rotation, scipy.optimize._direct, scipy.cluster._vq, scipy.cluster._hierarchy, scipy.cluster._optimal_leaf_ordering, scipy.ndimage._nd_image, _ni_label, scipy.ndimage._ni_label, scipy.special.cython_special, scipy.stats._stats, scipy.stats.beta_ufunc, scipy.stats._boost.beta_ufunc, scipy.stats.binom_ufunc, scipy.stats._boost.binom_ufunc, scipy.stats.nbinom_ufunc, scipy.stats._boost.nbinom_ufunc, scipy.stats.hypergeom_ufunc, scipy.stats._boost.hypergeom_ufunc, scipy.stats.ncf_ufunc, scipy.stats._boost.ncf_ufunc, scipy.stats.ncx2_ufunc, scipy.stats._boost.ncx2_ufunc, scipy.stats.nct_ufunc, scipy.stats._boost.nct_ufunc, scipy.stats.skewnorm_ufunc, scipy.stats._boost.skewnorm_ufunc, scipy.stats.invgauss_ufunc, scipy.stats._boost.invgauss_ufunc, scipy.interpolate._fitpack, scipy.interpolate.dfitpack, scipy.interpolate._bspl, scipy.interpolate._ppoly, scipy.interpolate.interpnd, scipy.interpolate._rbfinterp_pythran, scipy.interpolate._rgi_cython, scipy.stats._biasedurn, scipy.stats._levy_stable.levyst, scipy.stats._stats_pythran, scipy._lib._uarray._uarray, scipy.stats._ansari_swilk_statistics, scipy.stats._sobol, scipy.stats._qmc_cy, scipy.stats._mvn, scipy.stats._rcont.rcont, scipy.stats._unuran.unuran_wrapper, scipy.signal._sigtools, scipy.signal._max_len_seq_inner, scipy.signal._upfirdn_apply, scipy.signal._spline, scipy.signal._sosfilt, scipy.signal._spectral, scipy.signal._peak_finding_utils (total: 129)
python3.11-jax> ................[gw3] node down: Not properly terminated
```

zimbatm replied:
See https://gist.github.com/zimbatm/e9b61891f3bcf5e4aaefd13f94344fba

samuela replied:
Looking at the log I'm 95% confident this is an upstream issue.

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -77,7 +77,9 @@ let
     # https://github.com/NixOS/nixpkgs/pull/184395#issuecomment-1207287129
     # NOTE: We always build with NCCL; if it is unsupported, then our build is broken.
-    broken = effectiveStdenv.isDarwin || nccl.meta.unsupported;
+
+    # FIXME: Bazel fetch is not reproducible (see #321920).
+    broken = true;
```
Waiting for another collapse, but hopefully this is not needed anymore

zimbatm replied:
Ok, dropped

SomeoneSerge commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6114,13 +6114,14 @@ self: super: with self; {
   };
 
-  jaxlib = self.jaxlib-build;
+  # Using the -bin version since the -source one is currently broken (see #323154)
+  jaxlib = self.jaxlib-bin;
```
Personally, I've no preference as to whether this is implement in `jax/default.nix` or in `python-packages.nix`. Let's keep this as a source build on `linux-*`. Let's keep it at `-bin` for `darwin-*` but use `meta.unsupported` instead of `broken`

zimbatm replied:
Switched to this logic

SuperSandro2000 commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -130,10 +127,14 @@ buildPythonPackage rec {
+    # Segmentation fault. See https://gist.github.com/zimbatm/e9b61891f3bcf5e4aaefd13f94344fba
+    "tests/linalg_test.py"
+  ]
+  ++ (lib.optionals (stdenv.isDarwin && stdenv.isAarch64) [
```
```suggestion
  ++ lib.optionals (stdenv.isDarwin && stdenv.isAarch64) [
```

SuperSandro2000 commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -130,10 +127,14 @@ buildPythonPackage rec {
     # RuntimeWarning: invalid value encountered in cast
     "tests/lax_test.py"
-  ];
+  ]);
```
```suggestion
  ];
```

SomeoneSerge committed:
python311Packages.jaxlib: limit platforms

to the currently supported ones, so that the top-level jaxlib can fallback to jaxlib-bin

SomeoneSerge committed:
python3Packages.jax{lib,lib-bin}: disambiguate meta.descriptions

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -68,16 +68,17 @@ let
     # https://github.com/NixOS/nixpkgs/pull/184395#issuecomment-1207287129
-    # NOTE: We always build with NCCL; if it is unsupported, then our build is broken.
-    broken = effectiveStdenv.isDarwin || nccl.meta.unsupported;
+    platforms = platforms.linux;
```
This is to make the new `meta.unsupported` logic work. Honestly, I fail to predict what's this going to look like on search.nixos.org now

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -68,16 +68,17 @@ let
     # aarch64-darwin is broken because of https://github.com/bazelbuild/rules_cc/pull/136
     # however even with that fix applied, it doesn't work for everyone:
     # https://github.com/NixOS/nixpkgs/pull/184395#issuecomment-1207287129
-    # NOTE: We always build with NCCL; if it is unsupported, then our build is broken.
```
> # NOTE: We always build with NCCL; if it is unsupported, then our build is broken.

This is a peculiar comment, if the derivation depended on nccl and nccl was unsupported for the platform, the evaluation would fail saying exactly that: nccl is not supported on this platform. It also seems like the cpu-only jaxlib didn't actually have nccl in the inputs. Am I reading this right? CC @samuela

samuela replied:
i'm not entirely sure... my read is that `cudaSupport -> using nccl`, and the logic that was added just happened to forget about the CPU-only case.

cc @ConnorBaker who wrote this in https://github.com/NixOS/nixpkgs/commit/9bebd9e72d6b552fcfd3d1e6716eca6563944f42

SomeoneSerge committed:
python3Packages.jax: drop redundant parens

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>
